modules/OFConnectionManager2/module/src/ofconnectionmanager.c

```diff
@@ -0,0 +1,1774 @@
+controller_retry_ms(const controller_t *controller)
+{
+    if (controller->fail_count < 10) {
+        return 1 << controller->fail_count;
+    } else {
+        return 1000;
+    }
+}
+
+/* Note that cookie is the underlying controller of the main cxn */
+static void
+main_cxn_restart_handler(void *cookie)
+{
+    controller_t *controller = (controller_t *) cookie;
+    connection_t *cxn;
+
+    ind_soc_timer_event_unregister(main_cxn_restart_handler, cookie);
+
+    AIM_LOG_VERBOSE("Restarting %s(%p)", controller->desc, controller);
+
+    cxn = controller_add_cxn(controller, 0);
+    if (cxn == NULL) {
+        AIM_LOG_INTERNAL("Error allocating controller %s, "
+                         "rescheduling for %d ms", 
+                         controller->desc, controller_retry_ms(controller));
+        ind_soc_timer_event_register_with_priority(
+            main_cxn_restart_handler, controller,
+            controller_retry_ms(controller), IND_CXN_EVENT_PRIORITY);
+    }
+}
+
+static void
+cxn_closed_handler(void *cookie)
+{
+    connection_t *cxn = (connection_t *) cookie;
+    uint8_t aux_id = cxn->aux_id;
+    controller_t *controller = cxn->controller;
+
+    AIM_LOG_TRACE("%s: handling %s(%p)", __FUNCTION__, cxn->desc, cxn);
+
+    ind_cxn_free(cxn);
+    controller->cxns[aux_id] = NULL;
+
+    /* only restart the main connection */
+    if (aux_id == 0) {
+        if (controller->restartable) {
+            AIM_LOG_VERBOSE("Main cxn for controller %s rescheduled for %d ms",
+                            controller->desc, controller_retry_ms(controller));
+            ind_soc_timer_event_register_with_priority(
+                main_cxn_restart_handler, controller,
+                controller_retry_ms(controller), IND_CXN_EVENT_PRIORITY);
+        } else {
+            AIM_LOG_VERBOSE("Controller %s marked inactive",
+                            controller->desc);
+            controller->active = false;
+        }
+    }
+}
+
+/* called by cxn_instance to launch the cxn_closed_handler */
+void
+ind_cxn_notify_closed(connection_t *cxn)
+{
+    AIM_LOG_TRACE("%s: notifying for %s(%p), fail count %d",
+                  __FUNCTION__, cxn->desc, cxn,
+                  cxn->controller->fail_count);
+
+    /* use immediate timer to yield */
+    ind_soc_timer_event_register_with_priority(cxn_closed_handler, cxn,
+                                               IND_SOC_TIMER_IMMEDIATE,
+                                               IND_CXN_EVENT_PRIORITY);
+}
+
+
+/**
+ * Stop the specified connection if it is not yet closed.
+ * If it is closed, send a notification to clean up and restart.
+ */
+static void
+controller_stop_cxn(connection_t *cxn)
+{
+    if (cxn == NULL) {
+        AIM_LOG_INTERNAL("Controller stop on null cxn, skipping");
+        return;
+    }
+
+    AIM_LOG_TRACE("Controller stop cxn %s(%p)", cxn->desc, cxn);
+    if (!ind_cxn_is_closed(cxn)) {
+        ind_cxn_stop(cxn);
+    } else {
+        ind_cxn_notify_closed(cxn);
```

Comment:
for example, indigo_controller_remove() before we have gotten around to freeing the connection.